src/ray/core_worker/actor_manager.h

```diff
@@ -14,34 +14,122 @@
 
 #pragma once
 
+#include "absl/container/flat_hash_map.h"
 #include "ray/core_worker/actor_handle.h"
+#include "ray/core_worker/reference_count.h"
+#include "ray/core_worker/transport/direct_actor_transport.h"
 #include "ray/gcs/redis_gcs_client.h"
 
 namespace ray {
 
-// Interface for testing.
-class ActorManagerInterface {
- public:
-  virtual void PublishTerminatedActor(const TaskSpecification &actor_creation_task) = 0;
-
-  virtual ~ActorManagerInterface() {}
-};
-
 /// Class to manage lifetimes of actors that we create (actor children).
 /// Currently this class is only used to publish actor DEAD event
 /// for actor creation task failures. All other cases are managed
 /// by raylet.
-class ActorManager : public ActorManagerInterface {
+class ActorManager {
  public:
-  ActorManager(gcs::ActorInfoAccessor &actor_accessor)
-      : actor_accessor_(actor_accessor) {}
+  explicit ActorManager(
+      std::shared_ptr<gcs::GcsClient> gcs_client,
+      std::shared_ptr<CoreWorkerDirectActorTaskSubmitterInterface> direct_actor_submitter,
+      std::shared_ptr<ReferenceCounterInterface> reference_counter)
+      : gcs_client_(gcs_client),
+        direct_actor_submitter_(direct_actor_submitter),
+        reference_counter_(reference_counter) {}
+
+  ~ActorManager() {}
+
+  friend class ActorManagerTest;
+
+  /// Register an actor handle.
+  ///
+  /// This should be called when an actor handle is given to us by another task
+  /// or actor. This may be called even if we already have a handle to the same
+  /// actor.
+  ///
+  /// \param[in] actor_handle The actor handle.
+  /// \param[in] outer_object_id The object ID that contained the serialized
+  /// actor handle, if any.
+  /// \param[in] caller_id The caller's task ID
+  /// \param[in] call_site The caller's site.
+  /// \return The ActorID of the deserialized handle.
+  ActorID RegisterActorHandle(std::unique_ptr<ActorHandle> actor_handle,
+                              const ObjectID &outer_object_id, const TaskID &caller_id,
+                              const std::string &call_site,
+                              const rpc::Address &caller_address);
+
+  /// Get a handle to an actor.
+  ///
+  /// \param[in] actor_id The actor handle to get.
+  /// \return reference to the actor_handle's pointer.
+  /// NOTE: Returned actorHandle should not be stored anywhere.
+  std::unique_ptr<ActorHandle> &GetActorHandle(const ActorID &actor_id);
```

Comment:
Can the returned handle be `const`?

Reply:
Sure. Will do that!

Reply:
Actually, it doesn't seem to be unnecessary because this will make `unique_ptr` const. What we probably should do is to use `std::unique_ptr<const ActorHandle>` instead?

Reply:
We cannot use the `move` function if we make this function `const` because then the unique pointer should be const, which means it only allows copy operation. I will just keep as is. I tried making `ActorHandle` const, but it doesn't really make sense because we have setters for that class.

Reply:
Done